File: backend/src/main/java/org/akashihi/mdg/dao/AccountSpecification.java
package org.akashihi.mdg.dao;

import org.akashihi.mdg.entity.Account;
import org.akashihi.mdg.entity.AccountType;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.criteria.Predicate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;

public class AccountSpecification {
    public static Specification<Account> filteredAccount(Map<String, String> filter) {
        return (root, query, criteriaBuilder) -> {
            Collection<Predicate> predicates = new ArrayList<>();
            if (filter.containsKey("account_type")) {
                var at = AccountType.from(filter.get("account_type").toUpperCase());
                predicates.add(criteriaBuilder.equal(root.get("accountType"), at));
            }
            if (filter.containsKey("currency_id")) {
                predicates.add(criteriaBuilder.equal(root.get("currency.id"), Long.parseLong(filter.get("currency_id"))));
            }
            if (filter.containsKey("name")) {
                predicates.add(criteriaBuilder.equal(root.get("name"), filter.get("name")));
            }
            if (filter.containsKey("hidden")) {
                predicates.add(criteriaBuilder.equal(root.get("hidden"), Boolean.parseBoolean(filter.get("hidden"))));
            }
            return criteriaBuilder.and(predicates.toArray(new Predicate[predicates.size()]));
        };
    }
}

File: backend/src/main/kotlin/org/akashihi/mdg/dao/AccountSpecification.kt
package org.akashihi.mdg.dao

import org.akashihi.mdg.entity.Account
import org.akashihi.mdg.entity.AccountType.Companion.from
import org.springframework.data.jpa.domain.Specification
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.Predicate
import javax.persistence.criteria.Root
import kotlin.Any
import kotlin.String

object AccountSpecification {
    fun filteredAccount(filter: Map<String, String>): Specification<Account> {
        return Specification { root: Root<Account>, _: CriteriaQuery<*>, criteriaBuilder: CriteriaBuilder ->
            val predicates: MutableCollection<Predicate> = ArrayList()
            filter["account_type"]?.also { predicates.add(criteriaBuilder.equal(root.get<Any>("accountType"), from(it.uppercase()))) }
            filter["currency_id"]?.also { predicates.add(criteriaBuilder.equal(root.get<Any>("currency.id"), it.toLong())) }
            filter["name"]?.also { predicates.add(criteriaBuilder.equal(root.get<Any>("name"), it)) }
            filter["hidden"]?.also { predicates.add(criteriaBuilder.equal(root.get<Any>("hidden"), it.toBoolean())) }
            criteriaBuilder.and(*predicates.toTypedArray())
        }
    }
}
